src/groth16/offchain_checker.rs

```diff
@@ -0,0 +1,275 @@
+use crate::bn254::fp254impl::Fp254Impl;
+use crate::bn254::fq::Fq;
+
+use ark_ff::UniformRand;
+use ark_ff::{Field, One};
+use num_bigint::BigUint;
+use num_traits::{Num, ToPrimitive};
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+use std::str::FromStr;
+
+// refer table 3 of https://eprint.iacr.org/2009/457.pdf
+// a: Fp12 which is cubic residue
+// c: random Fp12 which is cubic non-residue
+// s: satisfying p^12 - 1 = 3^s * t
+// t: satisfying p^12 - 1 = 3^s * t
+// k: k = (t + 1) // 3
+fn tonelli_shanks_cubic(
+    a: ark_bn254::Fq12,
+    c: ark_bn254::Fq12,
+    s: u32,
+    t: BigUint,
+    k: BigUint,
+) -> ark_bn254::Fq12 {
+    let mut r = a.pow(t.to_u64_digits());
+    let e = 3_u32.pow(s - 1);
+    let exp = 3_u32.pow(s) * &t;
+
+    // compute cubic root of (a^t)^-1, say h
+    let (mut h, cc, mut c) = (
+        ark_bn254::Fq12::ONE,
+        c.pow([e as u64]),
+        c.inverse().unwrap(),
+    );
+    for i in 1..(s as i32) {
+        let delta = (s as i32) - i - 1;
+        let d = if delta < 0 {
+            r.pow((&exp / 3_u32.pow((-delta) as u32)).to_u64_digits())
+        } else {
+            r.pow([3_u32.pow(delta as u32).to_u64().unwrap()])
+        };
+        if d == cc {
+            (h, r) = (h * c, r * c.pow([3 as u64]));
+        } else if d == cc.pow([2_u64]) {
+            (h, r) = (h * c.pow([2 as u64]), r * c.pow([3 as u64]).pow([2 as u64]));
+        }
+        c = c.pow([3 as u64])
+    }
+
+    // recover cubic root of a
+    r = a.pow(k.to_u64_digits()) * h;
+    if t == 3_u32 * k + 1_u32 {
+        r = r.inverse().unwrap();
+    }
+
+    assert_eq!(r.pow([3 as u64]), a);
+    r
+}
+
+// refer from Algorithm 5 of "On Proving Pairings"(https://eprint.iacr.org/2024/640.pdf)
+pub fn compute_c_wi(f: ark_bn254::Fq12) -> (ark_bn254::Fq12, ark_bn254::Fq12) {
+    let p = BigUint::from_str_radix(Fq::MODULUS, 16).unwrap();
+    let r = BigUint::from_str(
+        "21888242871839275222246405745257275088548364400416034343698204186575808495617",
+    )
+    .unwrap();
+    let lambda = BigUint::from_str(
```

Comment:
you've already got lambda as a static Lazy in `src/groth16/constants.rs`. might want to grab it from there instead of re-parsing it everytime. consider moving other constants there as well

Reply:
yes, that would be better, if time permitted. thanks-)

Reply:
@rot13maxi is part of a group of token scammers who are trying to sabotage the BitVM project.